### PR TITLE
refactor: turn it into a class

### PR DESCRIPTION
**Pull request by gforcada**
A bigger step than #185 but on the same direction.

Move as much code as possible to functions.

This way, one can really see the flow/organization of the code much more clear.

The next step, in a future PR would be to create a class to group together all functions and minimize the global usage of variables, which would make it for a much nicer/easier to follow code base IMHO.

Is that something that would be welcomed? 🤔 🍀

### CODE REVIEW

**Comment by dataflake** [comment]
Making it easier to follow the code is always a good idea. @icemac what do you think?

**Comment by gforcada** [comment]
@dataflake @icemac I went ahead and refactored it once more to use a class, my final goal anyway 🍀 

Now, with `@cached_property` and avoiding that many argument passing at every function, the code looks much nicer IMHO ✨ 

I tried to be careful on copy&pasting the code around and adjusting, but we should double check that nothing got broken.

How does one test for that? 🤔 

If I run it against `zope.interface` from `master` branch and then again from this branch the diff should be almost zero, right? 🤔

File: config/config-package.py [hunk 1]
```diff
@@ -25,473 +26,620 @@
 FUTURE_PYTHON_VERSION = "3.12.0-alpha.2"
```
Comment (icemac) [inline config/config-package.py:26]:
```suggestion
FUTURE_PYTHON_VERSION = "3.12.0-alpha.2"
DEFAULT = object()
```

File: config/config-package.py [hunk 2]
```diff
@@ -25,473 +26,620 @@
...
+                self.config_type,
+                package_name=self.path.name,
+                setup=manylinux_install_setup,
+                aarch64_tests=manylinux_aarch64_tests,
+                with_future_python=self.with_future_python,
+            )
+            (self.path / '.manylinux-install.sh').chmod(0o755)
+            self.add_manylinux = True
+
+    def cfg_option(self, section, name, default=44):
```
Comment (icemac) [inline config/config-package.py:370]:
```suggestion
    def cfg_option(self, section, name, default=DEFAULT):
```

File: config/config-package.py [hunk 3]
```diff
@@ -25,473 +26,620 @@
...
+        """An empty list is not a good idea to have it as a default
+
+        Because it is mutable and it gets reused from one call to another.
+        """
+        if default == 44:
```
Comment (icemac) [inline config/config-package.py:375]:
```suggestion
        if default == DEFAULT:
```

File: config/config-package.py [hunk 4]
```diff
@@ -25,473 +26,620 @@
...
+                with_future_python=self.with_future_python,
+            )
+            (self.path / '.manylinux-install.sh').chmod(0o755)
+            self.add_manylinux = True
+
+    def cfg_option(self, section, name, default=44):
+        """An empty list is not a good idea to have it as a default
+
+        Because it is mutable and it gets reused from one call to another.
+        """
```
Comment (icemac) [inline config/config-package.py:374]:
```suggestion
        """Read a value from `self.meta_cfg`, default to `[]` if not existing.
        """
```

File: config/config-package.py [hunk 5]
```diff
@@ -25,473 +26,620 @@
...
+        if default == 44:
+            default = []
+        return self.meta_cfg[section].get(name, default)
+
+    def tox_option(self, name, default=44):
```
Comment (icemac) [inline config/config-package.py:379]:
```suggestion
    def tox_option(self, name, default=DEFAULT):
```

File: config/config-package.py [hunk 6]
```diff
@@ -25,473 +26,620 @@
...
+        """An empty list is not a good idea to have it as a default
+
+        Because it is mutable and it gets reused from one call to another.
+        """
```
Comment (icemac) [inline config/config-package.py:383]:
```suggestion
        """Read a value from the tox options.
        
        Default not existing ones to `[]`.
        """
```

File: config/config-package.py [hunk 7]
```diff
@@ -25,473 +26,620 @@
...
+        return self.cfg_option('tox', name, default)
+
+    def gh_option(self, name, default=44):
+        """An empty list is not a good idea to have it as a default
+
+        Because it is mutable and it gets reused from one call to another.
+        """
```
Comment (icemac) [inline config/config-package.py:390]:
```suggestion
    def gh_option(self, name, default=DEFAULT):
        """Read a value from the GitHub actions options.
        
        Default not existing ones to `[]`.
        """
```

File: config/config-package.py [hunk 8]
```diff
@@ -25,473 +26,620 @@
...
+            with_sphinx_doctests=self.with_sphinx_doctests,
+            with_future_python=self.with_future_python,
+            future_python_version=FUTURE_PYTHON_VERSION,
+            with_pypy=self.with_pypy,
+            with_macos=self.with_macos,
+            with_windows=self.with_windows,
+        )
+
+    def manifest_in(self):
+        """Modify MANIFEST.in with meta options"""
```
Comment (icemac) [inline config/config-package.py:489]:
```suggestion
        """Modify MANIFEST.in with meta options."""
```

**Review by icemac — APPROVED** [review]
I like the idea and added some suggestions to refactor the code a bit more.

**Comment by icemac** [comment]
> How does one test for that?

I always use a package where I try my changes and create a PR over there to show what calling that PR in meta/config results in.

**Comment by dataflake** [comment]
You could use some packages that I have configured only yesterday with the latest code from `master`, there should not be any changes except for maybe the `commit-id` value in `.meta.toml`:

- AccessControl
- Acquisition
- BTrees
- zope.security

All of those use the `c-code` template.

**Comment by gforcada** [comment]
@icemac thanks for all the feedback, all applied 👍🏾 

@dataflake thanks for the pointers, I will have a look right away 😃

**Comment by gforcada** [comment]
I did test the few packages that @dataflake suggested, and indeed, only the commit id on `.meta.cfg` was changed ✅ 

This should be ready to get merged 😃

**Review by dataflake — APPROVED** [review]

**Comment by dataflake** [comment]
You're free to merge when you have approved reviews, conversations are resolved and tests are green.

**Pull request merged**